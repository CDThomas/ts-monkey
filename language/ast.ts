export enum ASTKind {
  ArrayLiteral = "ARRAY_LITERAL",
  BlockStatement = "BLOCK_STATEMENT",
  Bool = "BOOL",
  CallExpression = "CALL_EXPRESSION",
  ExpressionStatement = "EXPRESSION",
  FunctionLiteral = "FUNCTION_LITERAL",
  HashLiteral = "HASH_LITERAL",
  Identifier = "IDENTIFIER",
  IfExpression = "IF_EXPRESSION",
  IndexExpression = "INDEX_EXPRESSION",
  InfixExpression = "INFIX_EXPRESSION",
  Integer = "INTEGER",
  Let = "LET",
  PrefixExpression = "PREFIX_EXPRESSION",
  Program = "PROGRAM",
  Return = "RETURN",
  String = "STRING"
}

export type Node = Program | Statement | Expression;

export type Statement =
  | BlockStatement
  | ExpressionStatement
  | LetStatement
  | ReturnStatment;

export type Expression =
  | ArrayLiteral
  | Bool
  | CallExpression
  | FunctionLiteral
  | HashLiteral
  | Identifier
  | IndexExpression
  | IfExpression
  | Integer
  | PrefixExpression
  | InfixExpression
  | Str;

export type Program = {
  kind: ASTKind.Program;
  statements: Statement[];
};

// Statements

export type BlockStatement = {
  kind: ASTKind.BlockStatement;
  statements: Statement[];
};

export type ExpressionStatement = {
  kind: ASTKind.ExpressionStatement;
  expression: Expression;
};

export type LetStatement = {
  kind: ASTKind.Let;
  name: Identifier;
  value: Expression;
};

export type ReturnStatment = {
  kind: ASTKind.Return;
  returnValue: Expression;
};

// Expressions

export type ArrayLiteral = {
  kind: ASTKind.ArrayLiteral;
  elements: Expression[];
};

export type Bool = {
  kind: ASTKind.Bool;
  value: boolean;
};

export type CallExpression = {
  kind: ASTKind.CallExpression;
  function: Identifier | FunctionLiteral;
  arguments: Expression[];
};

export type FunctionLiteral = {
  kind: ASTKind.FunctionLiteral;
  parameters: Identifier[];
  body: BlockStatement;
};

export type HashLiteral = {
  kind: ASTKind.HashLiteral;
  pairs: Map<Expression, Expression>;
};

export type Identifier = {
  kind: ASTKind.Identifier;
  value: string;
};

export type IfExpression = {
  kind: ASTKind.IfExpression;
  condition: Expression;
  consequence: BlockStatement;
  alternative?: BlockStatement;
};

export type IndexExpression = {
  kind: ASTKind.IndexExpression;
  left: Expression;
  index: Expression;
};

export type InfixExpression = {
  kind: ASTKind.InfixExpression;
  operator: string;
  left: Expression;
  right: Expression;
};

export type Integer = {
  kind: ASTKind.Integer;
  value: number;
};

export type PrefixExpression = {
  kind: ASTKind.PrefixExpression;
  operator: string;
  right: Expression;
};

export type Str = {
  kind: ASTKind.String;
  value: string;
};
