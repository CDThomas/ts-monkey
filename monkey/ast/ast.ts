import * as Token from "../token/token";

export type Node = {
  tokenLiteral(): string;
};

export type Statement = LetStatement;
export type Expression = Identifier;

class Program implements Node {
  statements: Statement[] = [];

  tokenLiteral(): string {
    if (this.statements.length > 0) {
      return this.statements[0].tokenLiteral();
    } else {
      return "";
    }
  }
}

class LetStatement implements Node {
  token: Token.Token; // the token.Let token
  name: Identifier;
  value?: Expression; // TODO: required

  constructor(token: Token.Token, name: Identifier, value?: Expression) {
    this.token = token;
    this.name = name;
    this.value = value;
  }

  tokenLiteral(): string {
    return this.token.literal;
  }
}

class Identifier implements Node {
  token: Token.Token; // the token.Ident token
  value: string;

  constructor(token: Token.Token, value: string) {
    this.token = token;
    this.value = value;
  }

  tokenLiteral(): string {
    return this.token.literal;
  }
}

export { LetStatement, Identifier, Program };
