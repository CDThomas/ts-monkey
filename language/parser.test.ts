import Parser from "./parser";
import Lexer from "./lexer";
import { ASTKind, Program } from "./ast";

function parse(input: string): Program {
  const lexer = new Lexer(input);
  const parser = new Parser(lexer);
  return parser.parseProgram();
}

describe("parsing", () => {
  test("let statements", () => {
    const input = `
    let x = 5;
    let y = 10;
    let foobar = 838383;
    `;

    const AST = parse(input);

    expect(AST).toEqual({
      kind: ASTKind.Program,
      statements: [
        { kind: ASTKind.Let, name: { kind: "IDENTIFIER", value: "x" } },
        { kind: ASTKind.Let, name: { kind: "IDENTIFIER", value: "y" } },
        { kind: ASTKind.Let, name: { kind: "IDENTIFIER", value: "foobar" } }
      ]
    });
  });

  test("return statements", () => {
    const input = `
    return 5;
    return 10;
    return 993322;
    `;

    const AST = parse(input);

    expect(AST).toEqual({
      kind: ASTKind.Program,
      statements: [
        { kind: ASTKind.Return },
        { kind: ASTKind.Return },
        { kind: ASTKind.Return }
      ]
    });
  });

  test("identifier expressions", () => {
    const input = "foobar";

    const AST = parse(input);

    expect(AST).toEqual({
      kind: ASTKind.Program,
      statements: [
        {
          kind: ASTKind.ExpressionStatement,
          expression: {
            kind: ASTKind.Identifier,
            value: "foobar"
          }
        }
      ]
    });
  });

  test("integer literal expressions", () => {
    const input = "5;";

    const AST = parse(input);

    expect(AST).toEqual({
      kind: ASTKind.Program,
      statements: [
        {
          kind: ASTKind.ExpressionStatement,
          expression: {
            kind: ASTKind.Integer,
            value: 5
          }
        }
      ]
    });
  });

  describe("prefix operators", () => {
    test("!", () => {
      const input = "!5";

      const AST = parse(input);

      expect(AST).toEqual({
        kind: ASTKind.Program,
        statements: [
          {
            kind: ASTKind.ExpressionStatement,
            expression: {
              kind: ASTKind.PrefixExpression,
              operator: "!",
              right: {
                kind: ASTKind.Integer,
                value: 5
              }
            }
          }
        ]
      });
    });

    test("-", () => {
      const input = "-15";

      const AST = parse(input);

      expect(AST).toEqual({
        kind: ASTKind.Program,
        statements: [
          {
            kind: ASTKind.ExpressionStatement,
            expression: {
              kind: ASTKind.PrefixExpression,
              operator: "-",
              right: {
                kind: ASTKind.Integer,
                value: 15
              }
            }
          }
        ]
      });
    });
  });
});
