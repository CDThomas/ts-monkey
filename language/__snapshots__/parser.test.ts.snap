// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsing operator precedence !-a 1`] = `
Object {
  "kind": "PROGRAM",
  "statements": Array [
    Object {
      "expression": Object {
        "kind": "PREFIX_EXPRESSION",
        "operator": "!",
        "right": Object {
          "kind": "PREFIX_EXPRESSION",
          "operator": "-",
          "right": Object {
            "kind": "IDENTIFIER",
            "value": "a",
          },
        },
      },
      "kind": "EXPRESSION",
    },
  ],
}
`;

exports[`parsing operator precedence -a * b 1`] = `
Object {
  "kind": "PROGRAM",
  "statements": Array [
    Object {
      "expression": Object {
        "kind": "INFIX_EXPRESSION",
        "left": Object {
          "kind": "PREFIX_EXPRESSION",
          "operator": "-",
          "right": Object {
            "kind": "IDENTIFIER",
            "value": "a",
          },
        },
        "operator": "*",
        "right": Object {
          "kind": "IDENTIFIER",
          "value": "b",
        },
      },
      "kind": "EXPRESSION",
    },
  ],
}
`;

exports[`parsing operator precedence 3 + 4; -5 * 5 1`] = `
Object {
  "kind": "PROGRAM",
  "statements": Array [
    Object {
      "expression": Object {
        "kind": "INFIX_EXPRESSION",
        "left": Object {
          "kind": "INTEGER",
          "value": 3,
        },
        "operator": "+",
        "right": Object {
          "kind": "INTEGER",
          "value": 4,
        },
      },
      "kind": "EXPRESSION",
    },
    Object {
      "expression": Object {
        "kind": "INFIX_EXPRESSION",
        "left": Object {
          "kind": "PREFIX_EXPRESSION",
          "operator": "-",
          "right": Object {
            "kind": "INTEGER",
            "value": 5,
          },
        },
        "operator": "*",
        "right": Object {
          "kind": "INTEGER",
          "value": 5,
        },
      },
      "kind": "EXPRESSION",
    },
  ],
}
`;

exports[`parsing operator precedence a * b * c 1`] = `
Object {
  "kind": "PROGRAM",
  "statements": Array [
    Object {
      "expression": Object {
        "kind": "INFIX_EXPRESSION",
        "left": Object {
          "kind": "INFIX_EXPRESSION",
          "left": Object {
            "kind": "IDENTIFIER",
            "value": "a",
          },
          "operator": "*",
          "right": Object {
            "kind": "IDENTIFIER",
            "value": "b",
          },
        },
        "operator": "*",
        "right": Object {
          "kind": "IDENTIFIER",
          "value": "c",
        },
      },
      "kind": "EXPRESSION",
    },
  ],
}
`;

exports[`parsing operator precedence a * b / c 1`] = `
Object {
  "kind": "PROGRAM",
  "statements": Array [
    Object {
      "expression": Object {
        "kind": "INFIX_EXPRESSION",
        "left": Object {
          "kind": "INFIX_EXPRESSION",
          "left": Object {
            "kind": "IDENTIFIER",
            "value": "a",
          },
          "operator": "*",
          "right": Object {
            "kind": "IDENTIFIER",
            "value": "b",
          },
        },
        "operator": "/",
        "right": Object {
          "kind": "IDENTIFIER",
          "value": "c",
        },
      },
      "kind": "EXPRESSION",
    },
  ],
}
`;

exports[`parsing operator precedence a + b * c + d / e - f 1`] = `
Object {
  "kind": "PROGRAM",
  "statements": Array [
    Object {
      "expression": Object {
        "kind": "INFIX_EXPRESSION",
        "left": Object {
          "kind": "INFIX_EXPRESSION",
          "left": Object {
            "kind": "INFIX_EXPRESSION",
            "left": Object {
              "kind": "IDENTIFIER",
              "value": "a",
            },
            "operator": "+",
            "right": Object {
              "kind": "INFIX_EXPRESSION",
              "left": Object {
                "kind": "IDENTIFIER",
                "value": "b",
              },
              "operator": "*",
              "right": Object {
                "kind": "IDENTIFIER",
                "value": "c",
              },
            },
          },
          "operator": "+",
          "right": Object {
            "kind": "INFIX_EXPRESSION",
            "left": Object {
              "kind": "IDENTIFIER",
              "value": "d",
            },
            "operator": "/",
            "right": Object {
              "kind": "IDENTIFIER",
              "value": "e",
            },
          },
        },
        "operator": "-",
        "right": Object {
          "kind": "IDENTIFIER",
          "value": "f",
        },
      },
      "kind": "EXPRESSION",
    },
  ],
}
`;

exports[`parsing operator precedence a + b + c 1`] = `
Object {
  "kind": "PROGRAM",
  "statements": Array [
    Object {
      "expression": Object {
        "kind": "INFIX_EXPRESSION",
        "left": Object {
          "kind": "INFIX_EXPRESSION",
          "left": Object {
            "kind": "IDENTIFIER",
            "value": "a",
          },
          "operator": "+",
          "right": Object {
            "kind": "IDENTIFIER",
            "value": "b",
          },
        },
        "operator": "+",
        "right": Object {
          "kind": "IDENTIFIER",
          "value": "c",
        },
      },
      "kind": "EXPRESSION",
    },
  ],
}
`;

exports[`parsing operator precedence a + b / c 1`] = `
Object {
  "kind": "PROGRAM",
  "statements": Array [
    Object {
      "expression": Object {
        "kind": "INFIX_EXPRESSION",
        "left": Object {
          "kind": "IDENTIFIER",
          "value": "a",
        },
        "operator": "+",
        "right": Object {
          "kind": "INFIX_EXPRESSION",
          "left": Object {
            "kind": "IDENTIFIER",
            "value": "b",
          },
          "operator": "/",
          "right": Object {
            "kind": "IDENTIFIER",
            "value": "c",
          },
        },
      },
      "kind": "EXPRESSION",
    },
  ],
}
`;

exports[`parsing operator precedence a + b - c 1`] = `
Object {
  "kind": "PROGRAM",
  "statements": Array [
    Object {
      "expression": Object {
        "kind": "INFIX_EXPRESSION",
        "left": Object {
          "kind": "INFIX_EXPRESSION",
          "left": Object {
            "kind": "IDENTIFIER",
            "value": "a",
          },
          "operator": "+",
          "right": Object {
            "kind": "IDENTIFIER",
            "value": "b",
          },
        },
        "operator": "-",
        "right": Object {
          "kind": "IDENTIFIER",
          "value": "c",
        },
      },
      "kind": "EXPRESSION",
    },
  ],
}
`;
